### PR TITLE
Monthly contribution flow

### PR DESCRIPTION
**Pull request by svillafe**
## Why are you doing this?
This code adds the new flow for creating new monthly contributions. 

**Before merging this PR, membership common should be updated:** https://github.com/guardian/membership-common/pull/435

## Changes
* The new product is under the URL /monthly-contributions
* The new URL is being added to robots.txt, therefore search engines can't crawl it.
* Added a new thank you page for the monthly contribution.
* If a user made a monthly contribution, he or she can't make a new one (it is being redirected to membership home).

## Flows
* If you are a monthly contributor, you can sign up for a membership.
* If you are a member, you can signup for a monthly contribution.
* If you are a monthly contributor, you can't set up another monthly contribution.
* If you are a member, you can't set up another membership.

## Future changes
Currently, the country is hardcoded, we should update the codebase in order to be able to manage currencies and different countries.

## Tests

| Browser | Supporter Stripe checkout | Supporter Paypal checkout | Stripe Monthly Contribution DEV |  Stripe Monthly Contribution UAT|
|---------|-----------------|-----------------|----------------------|----------------------|
| IE      |        ✅            |       ✅               |          ✅             |                     ✅    |
| Chrome  |   ✅                   |          ✅                 |        ✅                 | ✅                      |

### Step 1
mont1

### Step 2
month2

### Step 3
month3

### Step 4
month4

### Zuora DEV
picture 207

### CODE REVIEW

File: frontend/app/services/SalesforceService.scala [hunk 1]
```diff
@@ -51,21 +56,41 @@ class SalesforceService(salesforceConfig: SalesforceConfig) extends api.Salesfor
...
-      Keys.MAILING_CITY -> formData.deliveryAddress.town,
-      Keys.MAILING_STATE -> formData.deliveryAddress.countyOrState,
-      Keys.MAILING_POSTCODE -> formData.deliveryAddress.postCode,
-      Keys.MAILING_COUNTRY -> formData.deliveryAddress.country.getOrElse(UK).alpha2,
-      Keys.ALLOW_MEMBERSHIP_MAIL -> true
-    )) ++ Map(
-      Keys.ALLOW_THIRD_PARTY_EMAIL -> formData.marketingChoices.thirdParty,
-      Keys.ALLOW_GU_RELATED_MAIL -> formData.marketingChoices.gnm
-    ).collect { case (k, Some(v)) => Json.obj(k -> v) }
+  private def contributorData(customerOpt: Option[Customer]): JsObject =
```
Comment (jacobwinch) [inline frontend/app/services/SalesforceService.scala]: Do we need to store this information in Salesforce?

File: frontend/app/services/MemberService.scala [hunk 2]
```diff
@@ -184,6 +184,58 @@ class MemberService(identityService: IdentityService,
...
+        logger.error(s"Could not update Salesforce contact with membership status for user ${user.id}", e)
+      }
+
+
+    for {
+      stripeCustomer <- createStripeCustomer(formData.payment.stripeToken)
+      idUser <- getIdentityUserDetails
+      sfContact <- createSalesforceContact(idUser)
+      zuoraSubName <- createPaidZuoraSubscription(sfContact, formData, idUser.primaryEmailAddress, None, Some(stripeCustomer))
+      _ <- updateSalesforceContactWithContribution(Some(stripeCustomer)) // FIXME: This should go!
```
Comment (jacobwinch) [inline frontend/app/services/MemberService.scala]: For Supporter/Partner/Patron, I believe this call is only made to ensure that the membership tier is synced to Dynamo DB. (I'm not clear what extra info we would need to write directly to Salesforce for a monthly contribution?)

Consequently you might be able to remove this call altogether?

File: frontend/app/services/SalesforceService.scala (hunk 1 above)
Comment (jacobwinch) [inline frontend/app/services/SalesforceService.scala]: Looking more carefully, I can see that we write this information for membership too (although I think the tier update is the important part here). Generally speaking payment data only belongs in Zuora - I don't understand why we are writing it into Salesforce as well

File: frontend/app/services/MemberService.scala (hunk 2 above)
Comment (paulbrown1982) [inline frontend/app/services/MemberService.scala]: +1 we should not carry forward this monstrosity!

File: frontend/app/services/SalesforceService.scala (hunk 1 above)
Comment (paulbrown1982) [inline frontend/app/services/SalesforceService.scala]: +1 there's no reason to be sending payment information to Salesforce

File: frontend/app/controllers/Contributor.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  def joinMonthlyContribution = NonContributorAction.async { implicit request =>
+    monthlyContributorForm.bindFromRequest.fold({ formWithErrors =>
+      Future.successful(BadRequest(formWithErrors.errorsAsJson))
+    },
+      makeContributor(Ok(Json.obj("redirect" -> routes.Contributor.thankyouContributor().url))))
+  }
+
+  private def makeContributor(onSuccess: => Result)(formData: ContributorForm)(implicit request: AuthRequest[_]) = {
+    logger.info(s"User ${request.user.id} attempting to become a monthly contributor...")
+    val eventId = PreMembershipJoiningEventFromSessionExtractor.eventIdFrom(request.session)
```
Comment (paulbrown1982) [inline frontend/app/controllers/Contributor.scala]: This is not used.

File: frontend/app/controllers/Contributor.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    implicit val bp: BackendProvider = request
+    val idRequest = IdentityRequest(request)
+    val campaignCode = CampaignCode.fromRequest
+    val ipCountry = request.getFastlyCountry
+
+    Timing.record(salesforceService.metrics, "createMember") {
```
Comment (paulbrown1982) [inline frontend/app/controllers/Contributor.scala]: "createContributor"

File: frontend/app/controllers/Contributor.scala [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        // errors due to user's card are logged at WARN level as they are not logic errors
+        case error: Stripe.Error =>
+          logger.warn(s"Stripe API call returned error: \n\t$error \n\tuser=${request.user.id}")
+          Forbidden(Json.toJson(error))
+
+        case error: PaymentGatewayError =>
+          handlePaymentGatewayError(error, request.user.id, "monthly contributor", idRequest.trackingParameters)
+
+        case error =>
+          //salesforceService.metrics.putFailSignUp(tier)
```
Comment (jacobwinch) [inline frontend/app/controllers/Contributor.scala]: We can probably just delete this?

Comment (svillafe) [inline frontend/app/controllers/Contributor.scala]: 👍

File: frontend/app/forms/MemberForm.scala [hunk 6]
```diff
@@ -144,15 +168,14 @@ object MemberForm {
...
   private val productFeature = of[Set[FeatureChoice]] as productFeaturesFormatter
 
-
-  private val country: Mapping[String] =
+  private val country: Mapping[String] ={
     text.verifying { code => CountryGroup.countryByCode(code).isDefined }
       .transform(
         { code => CountryGroup.countryByCode(code).fold("")(_.name)},
         { name => CountryGroup.countryByNameOrCode(name).fold("")(_.alpha2)})
+  }
```
Comment (rtyley) [inline frontend/app/forms/MemberForm.scala]: Are these new brackets necessary?

File: frontend/app/views/support/ThankYouMonthlySummary.scala [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package views.support
+
+import com.gu.memsub.Price
+import org.joda.time.LocalDate
+
+case class ThankYouMonthlySummary(startDate: LocalDate,
+                                  amountPaidToday: Price) {
+
```
Comment (rtyley) [inline frontend/app/views/support/ThankYouMonthlySummary.scala]: Doesn't look like you need the `{ }` here!

**Comment by svillafe** [comment]
@rtyley you are right. I also see that there is a lot of repetition in the codebase. The main problem is that membership-frontend was conceived to have only a membership product. Introducing a new type of product, in this case contribution, it clashes with the existing codebase. I tried to keep the repetition to the minimum, but it is difficult without a global refactor of the project. As you said since we are going to build support soon, it is acceptable to merge this now.

**Comment by svillafe** [comment]
| Browser | Supporter Stripe checkout | Supporter Paypal checkout | Stripe Monthly Contribution DEV |  Stripe Monthly Contribution UAT|
|---------|-----------------|-----------------|----------------------|----------------------|
| IE      |                    |                      |                       |                        |
| Chrome  |        ✅               |                          |                      |                      |

**Comment by svillafe** [comment]
In the analytics, we can see people going to the thank you page after the deploy. 📈

**Pull request merged**